Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [ ]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

In [4]:


import pandas as pd
import requests
from bs4 import BeautifulSoup

import requests
import urllib.request
from urllib.request import urlopen
import time

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis

from pandas import DataFrame

In [3]:
pip install beautifulsoup4

     |████████████████████████████████| 122kB 21.8MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [5]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
response = requests.get(url)

In [16]:
soup = BeautifulSoup(response.content,'html.parser')


In [17]:
Zipcanada_df = soup.find_all('table')[0] 


In [6]:
Zipcanada_df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Take out all the rows from the Zipcanada table with column "Borough" not assigned.

In [4]:
Zip2 = Zipcanada_df[Zipcanada_df['Borough'] != 'Not assigned']

In [20]:
Zip2.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Check if there is "not assigned" value in the column: Neighborhood.
The logic is: if there is row like that, we will reassign it to equal to the value in borough. The result shows there is no value like that. So we do not have to do that.

In [5]:
Zip2.replace('Not assigned' , np.nan)
Zip2['Neighborhood'].isnull().values.any()

False

In [27]:
Zip2.shape


(103, 3)

The above are for the first part of the question. I got a table with 103 rows and 3 columns of zip code level data. 

Below are the second part of the question

Read in the CSV file for Zip code latitude and Longitute 

In [2]:
Geocanada_df = pd.read_csv('Geospatial_Coordinates.csv')


In [3]:
Geocanada_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [6]:
df_merge = pd.merge(Zip2, Geocanada_df, on='Postal Code')

In [7]:
df_merge.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

The above are for the second part of the question.


Now start the third part of the question and clean data for clustering analysis

Clean the data:
1. only work on the Borough with "Toronto" 
2. split the neighborhood if it has  more than 1 values.

In [8]:
Zip3_df =df_merge[df_merge['Borough'].str.contains("Toronto")]

In [9]:
Zip3_df[['N1','N2']] = Zip3_df["Neighborhood"].str.split(",", n = 1, expand = True)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/pandas/core/frame.py:2963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [10]:
Zip4 = Zip3_df.reset_index()

In [11]:
Zip4

index Postal Code           Borough  \
0       2         M5A  Downtown Toronto   
1       4         M7A  Downtown Toronto   
2       9         M5B  Downtown Toronto   
3      15         M5C  Downtown Toronto   
4      19         M4E      East Toronto   
5      20         M5E  Downtown Toronto   
6      24         M5G  Downtown Toronto   
7      25         M6G  Downtown Toronto   
8      30         M5H  Downtown Toronto   
9      31         M6H      West Toronto   
10     36         M5J  Downtown Toronto   
11     37         M6J      West Toronto   
12     41         M4K      East Toronto   
13     42         M5K  Downtown Toronto   
14     43         M6K      West Toronto   
15     47         M4L      East Toronto   
16     48         M5L  Downtown Toronto   
17     54         M4M      East Toronto   
18     61         M4N   Central Toronto   
19     62         M5N   Central Toronto   
20     67         M4P   Central Toronto   
21     68         M5P   Central Toronto   
22     69         M6P      West Toronto   
23     73         M4R   Central Toronto   
24     74         M5R   Central Toronto   
25     75         M6R      West Toronto   
26     79         M4S   Central Toronto   
27     80         M5S  Downtown Toronto   
28     81         M6S      West Toronto   
29     83         M4T   Central Toronto   
30     84         M5T  Downtown Toronto   
31     86         M4V   Central Toronto   
32     87         M5V  Downtown Toronto   
33     91         M4W  Downtown Toronto   
34     92         M5W  Downtown Toronto   
35     96         M4X  Downtown Toronto   
36     97         M5X  Downtown Toronto   
37     99         M4Y  Downtown Toronto   
38    100         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                      Lawrence Park  43.728020 -79.388790   
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
22                      High Park, The Junction South  43.661608 -79.464763   
23                  North Toronto West, Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
26                                         Davisville  43.704324 -79.388790   
27                     University of Toronto, Harbord  43.662696 -79.40

In [12]:
Zip5 = Zip4[['Postal Code','Borough', 'N1', 'Latitude', 'Longitude']]

In [13]:
Zip5

Postal Code           Borough                                     N1  \
0          M5A  Downtown Toronto                            Regent Park   
1          M7A  Downtown Toronto                           Queen's Park   
2          M5B  Downtown Toronto                        Garden District   
3          M5C  Downtown Toronto                         St. James Town   
4          M4E      East Toronto                            The Beaches   
5          M5E  Downtown Toronto                            Berczy Park   
6          M5G  Downtown Toronto                     Central Bay Street   
7          M6G  Downtown Toronto                               Christie   
8          M5H  Downtown Toronto                               Richmond   
9          M6H      West Toronto                               Dufferin   
10         M5J  Downtown Toronto                      Harbourfront East   
11         M6J      West Toronto                        Little Portugal   
12         M4K      East Toronto                      The Danforth West   
13         M5K  Downtown Toronto                Toronto Dominion Centre   
14         M6K      West Toronto                               Brockton   
15         M4L      East Toronto                           India Bazaar   
16         M5L  Downtown Toronto                         Commerce Court   
17         M4M      East Toronto                        Studio District   
18         M4N   Central Toronto                          Lawrence Park   
19         M5N   Central Toronto                               Roselawn   
20         M4P   Central Toronto                       Davisville North   
21         M5P   Central Toronto               Forest Hill North & West   
22         M6P      West Toronto                              High Park   
23         M4R   Central Toronto                     North Toronto West   
24         M5R   Central Toronto                              The Annex   
25         M6R      West Toronto                               Parkdale   
26         M4S   Central Toronto                             Davisville   
27         M5S  Downtown Toronto                  University of Toronto   
28         M6S      West Toronto                              Runnymede   
29         M4T   Central Toronto                             Moore Park   
30         M5T  Downtown Toronto                      Kensington Market   
31         M4V   Central Toronto                        Summerhill West   
32         M5V  Downtown Toronto                               CN Tower   
33         M4W  Downtown Toronto                               Rosedale   
34         M5W  Downtown Toronto                         Stn A PO Boxes   
35         M4X  Downtown Toronto                         St. James Town   
36         M5X  Downtown Toronto                   First Canadian Place   
37         M4Y  Downtown Toronto                   Church and Wellesley   
38         M7Y      East Toronto  Business reply mail Processing Centre   

     Latitude  Longitude  
0   43.654260 -79.360636  
1   43.662301 -79.389494  
2   43.657162 -79.378937  
3   43.651494 -79.375418  
4   43.676357 -79.293031  
5   43.644771 -79.373306  
6   43.657952 -79.387383  
7   43.669542 -79.422564  
8   43.650571 -79.384568  
9   43.669005 -79.442259  
10  43.640816 -79.381752  
11  43.647927 -79.419750  
12  43.679557 -79.352188  
13  43.647177 -79.381576  
14  43.636847 -79.428191  
15  43.668999 -79.315572  
16  43.648198 -79.379817  
17  43.659526 -79.340923  
18  43.728020 -79.388790  
19  43.711695 -79.416936  
20  43.712751 -79.390197  
21  43.696948 -79.411307  
22  43.661608 -79.464763  
23  43.715383 -79.405678  
24  43.672710 -79.405678  
25  43.648960 -79.456325  
26  43.704324 -79.388790  
27  43.662696 -79.400049  
28  43.651571 -79.484450  
29  43.689574 -79.383160  
30  43.653206 -79.400049  
31  43.686412 -79.400049  
32  43.628947 -79.394420  
33  43.679563 -79.377529  
34  43.646435 -79.374846  
35  43.667967 -79.367675  
36 

prepare the geo data for cluter 

In [14]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [15]:
# create map of New York using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(Zip3_df['Latitude'],Zip3_df['Longitude'], Zip3_df['Borough'], Zip3_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

use foursqaure API to get the venture information

In [16]:
CLIENT_ID = 'EI2DUPWAWDSGLRKX5W53NCHPOJ3JEK4KNJWRUL4YAX4VSDII' # your Foursquare ID
CLIENT_SECRET = 'CCDGYEKQYT2R0EZSSYEHFCLNS5PYKQZ20BVSZLPLKFM1LG0K' # your Foursquare Secret
VERSION = '20200718' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: EI2DUPWAWDSGLRKX5W53NCHPOJ3JEK4KNJWRUL4YAX4VSDII
CLIENT_SECRET:CCDGYEKQYT2R0EZSSYEHFCLNS5PYKQZ20BVSZLPLKFM1LG0K


In [17]:
LIMIT = 100 


radius = 500 

In [18]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [19]:
ca_venues = getNearbyVenues(names=Zip5['N1'],
                                   latitudes=Zip5['Latitude'],
                                   longitudes=Zip5['Longitude']
                                  )


Regent Park
Queen's Park
Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond
Dufferin
Harbourfront East
Little Portugal
The Danforth West
Toronto Dominion Centre
Brockton
India Bazaar
Commerce Court
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park
North Toronto West
The Annex
Parkdale
Davisville
University of Toronto
Runnymede
Moore Park
Kensington Market
Summerhill West
CN Tower
Rosedale
Stn A PO Boxes
St. James Town
First Canadian Place
Church and Wellesley
Business reply mail Processing Centre


In [20]:
print(ca_venues.shape)
ca_venues.head()

(1644, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park               43.65426              -79.360636   
1  Regent Park               43.65426              -79.360636   
2  Regent Park               43.65426              -79.360636   
3  Regent Park               43.65426              -79.360636   
4  Regent Park               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [21]:
ca_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                   
Berczy Park                                               59   
Brockton                                                  23   
Business reply mail Processing Centre                     19   
CN Tower                                                  17   
Central Bay Street                                        65   
Christie                                                  17   
Church and Wellesley                                      76   
Commerce Court                                           100   
Davisville                                                32   
Davisville North                                           9   
Dufferin                                                  13   
First Canadian Place                                     100   
Forest Hill North & West                                   5   
Garden District                                          100   
Harbourfront East                                        100   
High Park                                                 26   
India Bazaar                                              21   
Kensington Market                                         68   
Lawrence Park                                              4   
Little Portugal                                           47   
Moore Park                                                 1   
North Toronto West                                        20   
Parkdale                                                  14   
Queen's Park                                              32   
Regent Park                                               46   
Richmond                                                 100   
Rosedale                                                   4   
Roselawn                                                   2   
Runnymede                                                 35   
St. James Town                                           131   
Stn A PO Boxes                                            98   
Studio District                                           41   
Summerhill West                                           16   
The Annex                                                 21   
The Beaches                                                5   
The Danforth West                                         42   
Toronto Dominion Centre                                  100   
University of Toronto                                     35   

                                       Neighborhood Longitude  Venue  \
Neighborhood                                                           
Berczy Park                                                59     59   
Brockton                                                   23     23   
Business reply mail Processing Centre                      19     19   
CN Tower                                                   17     17   
Central Bay Street                                         65     65   
Christie                                                   17     17   
Church and Wellesley                                       76     76   
Commerce Court                                            100    100   
Davisville                                                 32     32   
Davisville North                                            9      9   
Dufferin                                                   13     13   
First Canadian Place                                      100    100   
Forest Hill North & West                                    5      5   
Garden District                                           100    100   
Harbourfront East                                         100    100   
High Park                                                  26     26   
India Bazaar                                               21     21   
Kensington Market                                          68     68   
Lawrence Park                         

In [22]:
print('There are {} uniques categories.'.format(len(ca_venues['Venue Category'].unique())))

There are 233 uniques categories.


In [23]:
# one hot encoding
ca_onehot = pd.get_dummies(ca_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ca_onehot['Neighborhood'] = ca_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [ca_onehot.columns[-1]] + list(ca_onehot.columns[:-1])
ca_onehot = ca_onehot[fixed_columns]

ca_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0

In [24]:
ca_onehot.shape

(1644, 233)

In [25]:
ca_grouped = ca_onehot.groupby('Neighborhood').mean().reset_index()
ca_grouped

Neighborhood  Yoga Studio  Afghan Restaurant  \
0                             Berczy Park     0.000000           0.000000   
1                                Brockton     0.043478           0.000000   
2   Business reply mail Processing Centre     0.000000           0.000000   
3                                CN Tower     0.000000           0.000000   
4                      Central Bay Street     0.015385           0.000000   
5                                Christie     0.000000           0.000000   
6                    Church and Wellesley     0.026316           0.013158   
7                          Commerce Court     0.000000           0.000000   
8                              Davisville     0.000000           0.000000   
9                        Davisville North     0.000000           0.000000   
10                               Dufferin     0.000000           0.000000   
11                   First Canadian Place     0.000000           0.000000   
12               Forest Hill North & West     0.000000           0.000000   
13                        Garden District     0.000000           0.000000   
14                      Harbourfront East     0.000000           0.000000   
15                              High Park     0.000000           0.000000   
16                           India Bazaar     0.000000           0.000000   
17                      Kensington Market     0.000000           0.000000   
18                          Lawrence Park     0.000000           0.000000   
19                        Little Portugal     0.021277           0.000000   
20                             Moore Park     0.000000           0.000000   
21                     North Toronto West     0.050000           0.000000   
22                               Parkdale     0.000000           0.000000   
23                           Queen's Park     0.031250           0.000000   
24                            Regent Park     0.021739           0.000000   
25                               Richmond     0.000000           0.000000   
26                               Rosedale     0.000000           0.000000   
27                               Roselawn     0.000000           0.000000   
28                              Runnymede     0.028571           0.000000   
29                         St. James Town     0.000000           0.000000   
30                         Stn A PO Boxes     0.010204           0.000000   
31                        Studio District     0.024390           0.000000   
32                        Summerhill West     0.000000           0.000000   
33                              The Annex     0.000000           0.000000   
34                            The Beaches     0.000000           0.000000   
35                      The Danforth West     0.023810           0.000000   
36                Toronto Dominion Centre     0.000000           0.000000   
37                  University of Toronto     0.028571           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.058824            0.058824      0.058824        0.117647   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000   

In [26]:
ca_grouped.shape

(38, 233)

In [27]:
num_top_venues = 5

for hood in ca_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = ca_grouped[ca_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.08
1  Cocktail Bar  0.03
2   Cheese Shop  0.03
3      Beer Bar  0.03
4          Café  0.03


----Brockton----
                venue  freq
0                Café  0.13
1         Coffee Shop  0.09
2      Breakfast Spot  0.09
3         Yoga Studio  0.04
4  Italian Restaurant  0.04


----Business reply mail Processing Centre----
                  venue  freq
0    Light Rail Station  0.11
1  Gym / Fitness Center  0.05
2         Auto Workshop  0.05
3            Comic Shop  0.05
4           Pizza Place  0.05


----CN Tower----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3             Plane  0.06
4          Boutique  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1   Italian Restaurant  0.06
2       Sandwich Place  0.05
3                 Café  0.05
4  Japanese Restaurant  0.05


----Christie----
           venue  freq
0  Grocery St

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = ca_grouped['Neighborhood']

for ind in np.arange(ca_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ca_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                            Berczy Park           Coffee Shop   
1                               Brockton                  Café   
2  Business reply mail Processing Centre    Light Rail Station   
3                               CN Tower      Airport Terminal   
4                     Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café           Cheese Shop        Farmers Market   
1        Breakfast Spot           Coffee Shop          Climbing Gym   
2                  Park         Auto Workshop            Comic Shop   
3        Airport Lounge       Airport Service                 Plane   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant            Restaurant   
1         Burrito Place    Italian Restaurant            Restaurant   
2           Pizza Place               Butcher      Recording Studio   
3       Harbor / Marina   Rental Car Location           Coffee Shop   
4   Japanese Restaurant           Salad Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar          Cocktail Bar                 Lounge  
1          Intersection               Stadium                    Bar  
2         Burrito Place            Restaurant                Brewery  
3         Boat or Ferry                   Bar           Airport Gate  
4          Burger Joint                   Bar       Department Store

Cluster by K means


In [33]:
# set number of clusters
kclusters = 3

ca_grouped_clustering = ca_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=2).fit(ca_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [34]:
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 2, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [35]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [36]:
neighborhoods_venues_sorted

Cluster Labels                           Neighborhood  \
0                1                            Berczy Park   
1                1                               Brockton   
2                1  Business reply mail Processing Centre   
3                1                               CN Tower   
4                1                     Central Bay Street   
5                1                               Christie   
6                1                   Church and Wellesley   
7                1                         Commerce Court   
8                1                             Davisville   
9                1                       Davisville North   
10               1                               Dufferin   
11               1                   First Canadian Place   
12               0               Forest Hill North & West   
13               1                        Garden District   
14               1                      Harbourfront East   
15               1                              High Park   
16               1                           India Bazaar   
17               1                      Kensington Market   
18               0                          Lawrence Park   
19               1                        Little Portugal   
20               2                             Moore Park   
21               1                     North Toronto West   
22               1                               Parkdale   
23               1                           Queen's Park   
24               1                            Regent Park   
25               1                               Richmond   
26               0                               Rosedale   
27               1                               Roselawn   
28               1                              Runnymede   
29               1                         St. James Town   
30               1                         Stn A PO Boxes   
31               1                        Studio District   
32               1                        Summerhill West   
33               1                              The Annex   
34               1                            The Beaches   
35               1                      The Danforth West   
36               1                Toronto Dominion Centre   
37               1                  University of Toronto   

   1st Most Common Venue          2nd Most Common Venue  \
0            Coffee Shop                           Café   
1                   Café                 Breakfast Spot   
2     Light Rail Station                           Park   
3       Airport Terminal                 Airport Lounge   
4            Coffee Shop             Italian Restaurant   
5          Grocery Store                           Café   
6            Coffee Shop            Japanese Restaurant   
7            Coffee Shop                           Café   
8         Sandwich Place                    Pizza Place   
9                   Park                            Gym   
10              Pharmacy                         Bakery   
11           Coffee Shop                           Café   
12                  Park                       Bus Line   
13        Clothing Store                    Coffee Shop   
14           Coffee Shop                       Aquarium   
15    Mexican Restaurant                           Café   
16                  Park                    Pizza Place   
17                  Café  Vegetarian / Vegan Restaurant   
18                  Park                    Swim School   
19                   Bar                     Restaurant   
20                 Trail                  Women's Store   
21           Coffee Shop                 Clothing Store   
22             Gift Shop                 Breakfast Spot   
23           Coffee Shop                          Diner   
24           Coffee Shop                           Park   
25           Coffee Shop                           Café   
26                  Park       

In [37]:
Zip6 = Zip5.merge(neighborhoods_venues_sorted, left_on='N1', right_on='Neighborhood')

In [46]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Zip6['Latitude'], Zip6['Longitude'], Zip6['Neighborhood'], Zip6['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

ca_merged = Zip5

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
ca_merged =ca_merged.join(neighborhoods_venues_sorted.set_index('N1'), on='Neighborhood')

ca_merged.head() # check the last columns!

ValueError: cannot insert Cluster Labels, already exists